[ao][fx] fixing rare bug with QuantizedGraphModule class

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89893
* #89892

when using PNP with functional linears (or other use cases with on the fly packing), the packedparams are created on the fly meaning the node.op is call_function rather than get_attr and node.target is a function rather than a str. This would cause the class to error unexpectedly since it expects a str.

Differential Revision: [D41578951](https://our.internmc.facebook.com/intern/diff/D41578951/)

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel

HDCharles committed:
[ao][fx] fixing rare bug with QuantizedGraphModule class

when using PNP with functional linears (or other use cases with on the fly packing), the packedparams are created on the fly meaning the node.op is call_function rather than get_attr and node.target is a function rather than a str. This would cause the class to error unexpectedly since it expects a str.

Differential Revision: [D41578951](https://our.internmc.facebook.com/intern/diff/D41578951/)

[ghstack-poisoned]

vkuzo commented:
should we add a test case?

Also, we discussed earlier that we want to delete the on-the-fly weight packing for performance reasons.  Is this fix still needed once that work is done?

HDCharles committed:
Update on "[ao][fx] fixing rare bug with QuantizedGraphModule class"


when using PNP with functional linears (or other use cases with on the fly packing), the packedparams are created on the fly meaning the node.op is call_function rather than get_attr and node.target is a function rather than a str. This would cause the class to error unexpectedly since it expects a str.

Differential Revision: [D41578951](https://our.internmc.facebook.com/intern/diff/D41578951/)

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][fx] fixing rare bug with QuantizedGraphModule class"


when using PNP with functional linears (or other use cases with on the fly packing), the packedparams are created on the fly meaning the node.op is call_function rather than get_attr and node.target is a function rather than a str. This would cause the class to error unexpectedly since it expects a str.

Differential Revision: [D41578951](https://our.internmc.facebook.com/intern/diff/D41578951/)

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][fx] fixing rare bug with QuantizedGraphModule class"


when using PNP with functional linears (or other use cases with on the fly packing), the packedparams are created on the fly meaning the node.op is call_function rather than get_attr and node.target is a function rather than a str. This would cause the class to error unexpectedly since it expects a str.

Differential Revision: [D41578951](https://our.internmc.facebook.com/intern/diff/D41578951/)

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][fx] fixing rare bug with QuantizedGraphModule class"


when using PNP with functional linears (or other use cases with on the fly packing), the packedparams are created on the fly meaning the node.op is call_function rather than get_attr and node.target is a function rather than a str. This would cause the class to error unexpectedly since it expects a str.

Differential Revision: [D41578951](https://our.internmc.facebook.com/intern/diff/D41578951/)

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][fx] fixing rare bug with QuantizedGraphModule class"


when using PNP with functional linears (or other use cases with on the fly packing), the packedparams are created on the fly meaning the node.op is call_function rather than get_attr and node.target is a function rather than a str. This would cause the class to error unexpectedly since it expects a str.

Differential Revision: [D41578951](https://our.internmc.facebook.com/intern/diff/D41578951/)

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/graph_module.py:
```diff
@@ -89,7 +89,11 @@ class QuantizedGraphModule(GraphModule):
     def __init__(self, root: Union[torch.nn.Module, Dict[str, Any]], graph: Graph, preserved_attr_names: Set[str]):
         self.preserved_attr_names = preserved_attr_names
-        preserved_attrs = {attr: getattr(root, attr) for attr in self.preserved_attr_names if hasattr(root, attr)}
+        preserved_attrs = {
```
when is `attr` not string?

jerryzh168 replied:
this is a bit weird since preserved_attr_names should be passed from users and it should be a set of `str` I think

HDCharles replied:
PNP has some weird cases where weight packing happens on the fly

jerryzh168 replied:
I see, should we just error out if user provides non str in `preserved_attr_names`?

HDCharles replied:
yeah, that's current behavior, I'm going to fix this elsewhere, I think you're right

HDCharles commented:
> should we add a test case?
> 
> Also, we discussed earlier that we want to delete the on-the-fly weight packing for performance reasons. Is this fix still needed once that work is done?

probably still want this fix for safety, just because PNP wont be doing on the fly packing doesn't mean it'll never come up.
